devicetypes/qubino/qubino-flush-dimmer.src/qubino-flush-dimmer.groovy

```diff
@@ -0,0 +1,890 @@
+	childDni = "${device.deviceNetworkId}-${childUniqueDni[componentLabel]}"
+
+	if (doesChildDeviceExist(componentLabel)) {
+		log.debug("Found device: ${componentLabel}")
+	} else {
+		log.debug ("Device ${componentLabel} not found. Creating new child device.")
+		createChildDevice(componentLabel, childDTHname, childDni)
+	}
+	sendEventToChild(map, childDni)
+}
+
+def sendEventToChild(event, childDni) {
+	def child = childDevices.find { it.deviceNetworkId == childDni }
+	log.debug "Sending event: ${event} to child: ${child}"
+	child?.sendEvent(event)
+}
+
+private refreshChild() {
+	sendHubCommand(zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 1, scale: 0).format())
+}
+
+def doesChildDeviceExist(componentLabel) {
+	def exists = (childDevices.find{it.toString() == componentLabel.toString()} == null) ? false : true
+}
+
+def getAppSwitchOnOffTime() {
+	/*
+	User can set on/off time between 50 - 255 ms only when the dimmer is activated from the physical wall-switch.
+	This function adjusts that time to be used also when activating from the Mobile App (1-3s).
+	*/
+	def dimmingTimePref = parameterMap.find({it.parameterNumber == 65})
+	def dimmingTimeBtn = getCommandValue(dimmingTimePref) ?: dimmingTimePref.defaultValue
+	return Math.round(1 + ((dimmingTimeBtn - 50) * (3-1))/ (255-50))
+}
+
+def on() {
+	def maxDimmingLvlPref = parameterMap.find({it.parameterNumber == 61})
+	def value = state.lastDimmingLevel ?: getCommandValue(maxDimmingLvlPref) ?: maxDimmingLvlPref.defaultValue
+	def dimmingOn = appSwitchOnOffTime
+
+	log.debug ("dimming on value: ${value}")
+	log.debug ("dimming on time: ${dimmingOn}")
+
+	encapSequence([
+			zwave.switchMultilevelV3.switchMultilevelSet(dimmingDuration: dimmingOn, value: value),
+			//zwave.switchMultilevelV3.switchMultilevelGet(),
+			meterGet(scale: 0),
+			meterGet(scale: 2)
+	], 5000)
+}
+
+def off() {
+	def dimmingOff = appSwitchOnOffTime
+	log.debug ("dimming off time: ${dimmingOff}")
+	encapSequence([
+			zwave.switchMultilevelV3.switchMultilevelSet(dimmingDuration: dimmingOff, value: 0x00),
+			//zwave.switchMultilevelV3.switchMultilevelGet(),
+			meterGet(scale: 0),
+			meterGet(scale: 2)
+	], 5000)
+}
+
+def ping() {
+	refresh()
+}
+
+def refresh() {
+	log.debug "refresh()"
+	encapSequence([
+			zwave.switchMultilevelV1.switchMultilevelGet(),
+			meterGet(scale: 0),
+			meterGet(scale: 2),
+			zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 1, scale: 0)
+	], 100)
+}
+
+def checkIfLevelCorrect(level) {
+	def max = getCommandValue(parameterMap.find({it.parameterNumber == 61})) ?: maxDimmingLvlPref.defaultValue
+	def min = getCommandValue(parameterMap.find({it.parameterNumber == 60})) ?: minDimmingLvlPref.defaultValue
+
+	if (level > max && level <= 100) {
+		level = max
+		log.warn "setLvl is greater than ${max}. New lvl: ${level}"
+	} else if (level < min) {
+		level = min
+		log.warn "setLvl is lower than ${min}. New lvl: ${level}"
+	}
+	return level as int
+}
+
+def setLevel(level) {

```

Comment:
the signature needs to include the `rate` argument as well for this to work correctly with smart apps. I don't know how you want to make that work with the preference though, probably just discard it and favor the preference value